graphbolt/src/cuda/neighbor_sampler.cu

```diff
@@ -123,24 +124,112 @@ struct SegmentEndFunc {
   }
 };
 
+template <typename indptr_t, typename etype_t>
+struct EdgeTypeSearch {
+  const indptr_t* sub_indptr;
+  const indptr_t* sliced_indptr;
+  const etype_t* etypes;
+  int64_t num_fanouts;
+  int64_t num_rows;
+  indptr_t* new_sub_indptr;
+  indptr_t* new_sliced_indptr;
+  __host__ __device__ auto operator()(int64_t i) {
+    const auto homo_i = i / num_fanouts;
+    const auto indptr_i = sub_indptr[homo_i];
+    const auto degree = sub_indptr[homo_i + 1] - indptr_i;
+    const etype_t etype = i % num_fanouts;
+    auto offset = cuda::LowerBound(etypes + indptr_i, degree, etype);
+    new_sub_indptr[i] = indptr_i + offset;
+    new_sliced_indptr[i] = sliced_indptr[homo_i] + offset;
+    if (i == num_rows - 1) new_sub_indptr[num_rows] = indptr_i + degree;
+  }
+};
+
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> SliceCSCIndptrHetero(
```

Comment:
why not place it together with `SliceCSCIndptr` in `index_select_csc_impl.cu`? or is it possible to merge them into unified one?

Reply:
Maybe we can take them out to a different file together?

Reply:
yes, utility functions is better.

Reply:
even in another separate PR works for me.

Reply:
It would be hard for me to test the correctness of the functionality in 2 separate PRs. If you don't mind, could we review the changes together? I moved the SliceCSCIndptr function from one file to another so we don't have to check it at all.